### PR TITLE
add cluster-machine-approver

### PR DESCRIPTION
**Pull request by hanqiuzh**
Updates:
- removed default auto-approver which approve all CSRs
- used kube-controller-manager to approve kubelet client CSR
- used cluster-machine-approver to approve kubelet serving CSR

### CODE REVIEW

**Comment by hanqiuzh** [comment]
/hold
currently using a non-ocp image of cluster-machine-approver, will need to update with a more formal image

File: control-plane-operator/controllers/hostedcontrolplane/assets/cluster-machine-approver/cluster-machine-approver-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      labels:
+        app: machine-approver
+    spec:
+      serviceAccountName: machine-approver-sa
+      containers:
+      - name: machine-approver-controller
+{{- if atleast_version "4.10.0" }}
+        image: {{ imageFor "cluster-machine-approver" }}
+{{- else }}
+        image: quay.io/hanqiuzh/origin-cluster-machine-approver@sha256:ae555a2d6bbf99ea5ed0764216f06c759ac0fd861340afda8f714a29bd2da44b
```
Comment (hanqiuzh) [inline control-plane-operator/controllers/hostedcontrolplane/assets/cluster-machine-approver/cluster-machine-approver-deployment.yaml:28]: cannot find an latest stable image we can reference, so used one built from source

File: control-plane-operator/controllers/hostedcontrolplane/render/manifests.go [hunk 2]
```diff
@@ -64,6 +64,7 @@ func (c *clusterManifestContext) setupManifests() {
 	c.userManifestsBootstrapper()
 	c.machineConfigServer()
 	c.ignitionConfigs()
+	c.clusterMachineApprover()
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/render/manifests.go:67]: The Machine Approver operate with CAPI resources. These are not HCP versioned. Instead they have independent release cycles. The particular version for capi, autoscaler and machine approver should dictated and reconciled by the Hypershift HostedCluster controller. See
https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#L78-L84

Comment (hanqiuzh) [reply control-plane-operator/controllers/hostedcontrolplane/render/manifests.go:67]: sure, that makes sense. will move the approver into the hostedcluster controller. 👍

Comment (hanqiuzh) [reply control-plane-operator/controllers/hostedcontrolplane/render/manifests.go]: updated to move everything into the hostedcluster controller now 👍

File: control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/cluster-csr-approver-clusterrole.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
+# https://github.com/openshift/cluster-kube-controller-manager-operator/blob/release-4.9/bindata/assets/kube-controller-manager/csr_approver_clusterrole.yaml
```
Comment (hanqiuzh) [inline control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/cluster-csr-approver-clusterrole.yaml:1]: this clusterrole is used for cluster-policy-controller to approve monitoring csr

File: control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/node-bootstrapper-csr-approval-rolebinding.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Approve all CSRs for the group "system:bootstrappers"
```
Comment (hanqiuzh) [inline control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/node-bootstrapper-csr-approval-rolebinding.yaml:1]: this clusterrolebinding will let kube-controller-manager approve kubelet client csr. reference: https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet-tls-bootstrapping/#approval

File: control-plane-operator/controllers/hostedcontrolplane/clusterpolicy/deployment.go [hunk 5]
```diff
@@ -70,6 +70,12 @@ func buildOCMContainerMain(image string) func(*corev1.Container) {
 			"--namespace=openshift-kube-controller-manager",
 		}
 		c.VolumeMounts = volumeMounts.ContainerMounts(c.Name)
+		c.Env = []corev1.EnvVar{
```
Comment (hanqiuzh) [inline control-plane-operator/controllers/hostedcontrolplane/clusterpolicy/deployment.go:73]: adding this env var to allow policy-controller create events in `openshift-kube-controller-manager` namespace

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 6]
```diff
@@ -82,6 +84,9 @@ const (
 	imageCAPI = "gcr.io/k8s-staging-cluster-api/cluster-api-controller@sha256:9288d085f9fa52bb7186e0de2971945cd192d3a1e849135f7381e4f08fe28442"
 	// This comes from https://console.cloud.google.com/gcr/images/k8s-artifacts-prod
 	imageCAPA = "us.gcr.io/k8s-artifacts-prod/cluster-api-aws/cluster-api-aws-controller:v0.7.0"
+	// This image is built from https://github.com/openshift/cluster-machine-approver/tree/release-4.10
+	imageMachineApprover        = "quay.io/openshift/origin-cluster-machine-approver:4.10.0"
```
Comment (hanqiuzh) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:88]: using upstream 4.10.0 image of cluster-machine-approver

**Comment by hanqiuzh** [comment]
@enxebre please review again 👍  now the upstream 4.10.0 machine-approver image is available

File: control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/node-bootstrapper-csr-approval-rolebinding.yaml (hunk 4 above)
Comment (enxebre) [reply control-plane-operator/controllers/hostedcontrolplane/assets/cluster-bootstrap/node-bootstrapper-csr-approval-rolebinding.yaml:1]: can you add this comment as a comment in the file?

File: control-plane-operator/controllers/hostedcontrolplane/clusterpolicy/deployment.go (hunk 5 above)
Comment (enxebre) [reply control-plane-operator/controllers/hostedcontrolplane/clusterpolicy/deployment.go:73]: can you add this comment as a comment in code?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 7]
```diff
@@ -2732,3 +2741,228 @@ func reconcileMachineConfigServerService(svc *corev1.Service) error {
...
+			Namespace: sa.Namespace,
+		},
+	}
+	return nil
+}
+
+func reconcileMachineApproverDeployment(deployment *appsv1.Deployment, hc *hyperv1.HostedCluster, sa *corev1.ServiceAccount, kubeConfigSecret *corev1.Secret, config *corev1.ConfigMap, machineApproverImage, machineApproverVersion string) error {
+	args := []string{
+		"--config=/var/run/configmaps/config/config.yaml",
+		"-v=4",
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2877]: we usually use v=3.

Can we enable leader election here?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 8]
```diff
@@ -2732,3 +2741,228 @@ func reconcileMachineConfigServerService(svc *corev1.Service) error {
...
+							},
+							{
+								Name:      "config",
+								MountPath: "/var/run/configmaps/config",
+							},
+						},
+						Env: []corev1.EnvVar{
+							{
+								Name:  "RELEASE_VERSION",
+								Value: machineApproverVersion,
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2946]: what's this environment variable for?

File: api/v1alpha1/hostedcluster_types.go [hunk 9]
```diff
@@ -57,6 +57,10 @@ const (
 	// AWSCredentialsFileSecretKey defines the Kubernetes secret key name that contains
 	// the customer AWS credentials in the unmanaged authentication strategy for AWS KMS secret encryption
 	AWSCredentialsFileSecretKey = "credentials"
+	// MachineApproverImage is an annotation that allows the specification of the machine approver image.
+	// This is a temporary workaround necessary for compliance reasons on the IBM Cloud side:
+	// no images can be pulled from registries outside of IBM Cloud's official regional registries
+	MachineApproverImage = "hypershift.openshift.io/machine-approver-image"
```
Comment (enxebre) [inline api/v1alpha1/hostedcluster_types.go:63]: cc @relyt0925 FYI.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 10]
```diff
@@ -82,6 +84,9 @@ const (
 	imageCAPI = "gcr.io/k8s-staging-cluster-api/cluster-api-controller@sha256:9288d085f9fa52bb7186e0de2971945cd192d3a1e849135f7381e4f08fe28442"
 	// This comes from https://console.cloud.google.com/gcr/images/k8s-artifacts-prod
 	imageCAPA = "us.gcr.io/k8s-artifacts-prod/cluster-api-aws/cluster-api-aws-controller:v0.7.0"
+	// This image is built from https://github.com/openshift/cluster-machine-approver/tree/release-4.10
+	imageMachineApprover        = "quay.io/openshift/origin-cluster-machine-approver:4.10.0"
+	imageMachineApproverVersion = "4.10.0"
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:89]: why do we need the version to be broken down?

**Comment by enxebre** [comment]
thanks @hanqiuzh! this is looking great! I dropped some comments.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 8 above)
Comment (hanqiuzh) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2946]: this will be showed as ClusterOperator CR's version. (`oc get co`)

Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2946]: I don't think we need to set this nor break the version down from the image name. We are not even deploying the clusterOperator for the machine approver right? The version here is dictated by the Hypershift release.

Comment (hanqiuzh) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2946]: good question, let me try again without setting this env. machine-approver was writing the clusteroperator's status with a version, before I used my customized image, and it was empty, let me try again 👍

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 11]
```diff
@@ -2732,3 +2741,228 @@ func reconcileMachineConfigServerService(svc *corev1.Service) error {
...
+
+type ClusterMachineApproverConfig struct {
+	NodeClientCert NodeClientCert `json:"nodeClientCert,omitempty"`
+}
+type NodeClientCert struct {
+	Disabled bool `json:"disabled,omitempty"`
+}
+
+func reconcileMachineApproverConfig(cm *corev1.ConfigMap) error {
+	// disable the client cert csr approval
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2823]: why?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 8 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2946]: >machine-approver was writing the clusteroperator's status with a version

The clusterOperator is a cluster scope resource so we can not deploy it here by design. This PR is not even making that possible because is not given clusterOperator rbac which is fine. My expectation would be for the status controller to just being unable to operate because the lack of rbac.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 11 above)
Comment (hanqiuzh) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2823]: Used kube-controller-manager for client CSR in this PR.

The machine-approver [client csr approval](https://github.com/openshift/cluster-machine-approver#node-client-csr-approval-workflow) requires a lot of hard coded values (from service account related to machine-config-operator?) like the following:
- The user must be the node bootstrapper
   - The username in the CSR must be `system:serviceaccount:openshift-machine-config-operator:node-bootstrapper`
   - The groups in the CSR must be `system:serviceaccounts:openshift-machine-config-operator`, `system:serviceaccounts`, and `system:authenticated`.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 8 above)
Comment (hanqiuzh) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2946]: oh, for hosted cluster I gave the `service-network-admin-kubeconfig`, so machine-approver has admin permission, and it will be able to update clusteroperator CR on hosted cluster

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 7 above)
Comment (hanqiuzh) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2877]: Didn't find any leader election flags in current machine-approver 😢

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 8 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2946]: >oh, for hosted cluster I gave the service-network-admin-kubeconfig

can you please point me to where in this PR is that happening?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 12]
```diff
@@ -2732,3 +2741,228 @@ func reconcileMachineConfigServerService(svc *corev1.Service) error {
...
+
+	// The deployment depends on the kubeconfig being reported.
+	if hcp.Status.KubeConfig != nil {
+		// Resolve the kubeconfig secret for machine-approver
+		kubeconfigSecret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: hcp.Namespace,
+				Name:      "service-network-admin-kubeconfig",
+			},
+		}
```
Comment (hanqiuzh) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2792]: @enxebre  for hosted cluster's kubeconfig, I used `service-network-admin-kubeconfig`, which I copied from other components 😂 

it will be mounted onto machine-approver, and be referenced as `"--workload-cluster-kubeconfig=/etc/kubernetes/kubeconfig/kubeconfig"`

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 8 above)
Comment (hanqiuzh) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2946]: yes, see https://github.com/openshift/hypershift/pull/477#discussion_r716848971

Comment (hanqiuzh) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2946]: Tried upstream 4.10.0 image without the RELEASE_VERSION set, and it will show like the following:
```
# oc get co
NAME                                       VERSION      AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
kube-apiserver                             4.9.0-rc.2   True        False         False      2m16s   
kube-controller-manager                    4.9.0-rc.2   True        False         False      2m16s   
...
machine-approver                           unknown      True        False         False      2m50s  
```

Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2947]: ># oc get co

Is this in the guest cluster?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 12 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2792]: ok make sense.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 8 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2947]: So the fact that the clusteroperator is created in the guest cluster and that csr and nodes details might be kube version specific would make me reconsider to run the machine approver from the payload by the HCP instead. Though I'd rather avoid that at least in the short term so we don't need to deal with backporting the machine-approver capi support.

So for now in this PR we have 3 options:
- Leave the clusteroperator version unknown
- Set the clusteroperator version, but it will always show 4.10 even for 4.8 HC.
- Add support in the mnachine approver to not run the cluster operator status controller, so it does not get deployed.

My preference would be the first one.
cc @csrwng, @ironcladlou @sjenning

Comment (hanqiuzh) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go]: yes, `oc get co` is inside guest cluster. 

I updated to use the first option, and it will show version=unknown for now 👍

Comment (csrwng) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go]: @enxebre for now I think it's fine to go with option 1, although if possible I'd like to eventually implement option 3. There is no part of the approver that is running in the guest cluster.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 10 above)
Comment (hanqiuzh) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go]: see https://github.com/openshift/hypershift/pull/477#discussion_r717615659 . updated to don't use version, and will show version=unknown when user run `oc get co` in guest cluster

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 7 above)
Comment (hanqiuzh) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go]: added a TODO comment for leader election

**Comment by hanqiuzh** [comment]
@enxebre @csrwng thanks for review. 
Based on the comments, I think this PR is basically ready? anything else in your mind I should update? please let me know 👍

**Comment by enxebre** [comment]
thanks @hanqiuzh, needs to investigate this before merging https://github.com/openshift/hypershift/pull/477#discussion_r716823690

**Comment by enxebre** [comment]
As a hard requirement we need to get client certificates through the machine approver.
This is where the client cert for the bootstrap-kubconfig is being generated https://github.com/openshift/hypershift/blob/a86c74d2528991839e09725660b56dfa5fddbca3/control-plane-operator/controllers/hostedcontrolplane/pki/kas.go#L52. So this is what we need to adapt to satisfy the machine approver.

We then generate the bootstrap-kubeconfig https://github.com/openshift/hypershift/blob/b427cf5dee50a794349f49a59551c49acdee0f90/control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go#L1458-L1464 and pass it through the MCS pod.

FWIW This is how the machine config generates the bootrstrap-kubconfig - https://github.com/openshift/machine-config-operator/blob/9c6c2bfd7ed498bfbc296d530d1839bd6a177b0b/pkg/server/cluster_server.go#L116

cc @hanqiuzh @csrwng

**Comment by csrwng** [comment]
@hanqiuzh The e2e failed because:
```
util.go:349: Container machine-approver-controller in pod machine-approver-6558676548-5spw9 has a restartCount > 0 (2)
```
W1006 02:34:01.303036       1 client_config.go:615] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
F1006 02:34:02.435053       1 main.go:85] Can't create clients: failed to create client: Get "https://kube-apiserver:6443/api?timeout=32s": dial tcp 172.30.183.11:6443: connect: connection refused
goroutine 1 [running]:
k8s.io/klog/v2.stacks(0xc000012001, 0xc000358000, 0xc5, 0xe2)
	/go/src/github.com/openshift/cluster-machine-approver/vendor/k8s.io/klog/v2/klog.go:1026 +0xb9
k8s.io/klog/v2.(*loggingT).output(0x2b6d100, 0xc000000003, 0x0, 0x0, 0xc000495f80, 0x0, 0x22f4876, 0x7, 0x55, 0x0)
	/go/src/github.com/openshift/cluster-machine-approver/vendor/k8s.io/klog/v2/klog.go:975 +0x1e5
k8s.io/klog/v2.(*loggingT).printf(0x2b6d100, 0xc000000003, 0x0, 0x0, 0x0, 0x0, 0x1c688b6, 0x18, 0xc0005143d0, 0x1, ...)
	/go/src/github.com/openshift/cluster-machine-approver/vendor/k8s.io/klog/v2/klog.go:753 +0x19a
k8s.io/klog/v2.Fatalf(...)
	/go/src/github.com/openshift/cluster-machine-approver/vendor/k8s.io/klog/v2/klog.go:1514
main.main()
	/go/src/github.com/openshift/cluster-machine-approver/main.go:85 +0x706
```
You may want to add an init container to the approver pod so that it waits for the kubeapiserver to be available. As in
https://github.com/openshift/hypershift/pull/517

cc @alvaroaleman

File: control-plane-operator/controllers/hostedcontrolplane/nodebootstrapper/kubeconfig.go [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+		secret.Labels = map[string]string{}
+	}
+	//secret.Labels[kas.KubeconfigScopeLabel] = string(scope)
+	secret.Labels["hypershift.openshift.io/kubeconfig"] = "bootstrap"
+	secret.Data[key] = kcData
+	return nil
+}
+
+func kubeconfigFromSecret(secret *corev1.Secret, apiserverURL string) ([]byte, []byte, error) {
+	caData := secret.Data[corev1.ServiceAccountRootCAKey]
```
Comment (hanqiuzh) [inline control-plane-operator/controllers/hostedcontrolplane/nodebootstrapper/kubeconfig.go:164]: I used the ca from service account directly (same as the machine-config-operator), but I guess I can also get ca from root-ca secret? not sure which is better

File: control-plane-operator/controllers/hostedcontrolplane/nodebootstrapper/kubeconfig.go [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+	if secret.Labels == nil {
+		secret.Labels = map[string]string{}
+	}
+	//secret.Labels[kas.KubeconfigScopeLabel] = string(scope)
+	secret.Labels["hypershift.openshift.io/kubeconfig"] = "bootstrap"
+	secret.Data[key] = kcData
+	return nil
+}
+
+func kubeconfigFromSecret(secret *corev1.Secret, apiserverURL string) ([]byte, []byte, error) {
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/nodebootstrapper/kubeconfig.go:163]: there's logic that does this in kas as it was before, why are duplicating this?

File: control-plane-operator/controllers/hostedcontrolplane/nodebootstrapper/kubeconfig.go [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+		}
+	}
+	if config == nil {
+		return fmt.Errorf("failed to load hosted controlplane kubeconfig from secret %q: %w", kubeconfigSecret.Name, err)
+	}
+	clientConfig, err := clientcmd.NewDefaultClientConfig(*config, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return fmt.Errorf("failed to create hosted controlplane client from secret %q: %w", kubeconfigSecret.Name, err)
+	}
+	hostedClusterClient, err := client.New(clientConfig, client.Options{})
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/nodebootstrapper/kubeconfig.go:62]: why do we need to do anything in the guest cluster? Isn't all we need to adjust https://github.com/openshift/hypershift/blob/a86c74d2528991839e09725660b56dfa5fddbca3/control-plane-operator/controllers/hostedcontrolplane/pki/kas.go#L52?

**Comment by enxebre** [comment]
@hanqiuzh I'm confused after latest changes, isn't all we need to adjust https://github.com/openshift/hypershift/blob/a86c74d2528991839e09725660b56dfa5fddbca3/control-plane-operator/controllers/hostedcontrolplane/pki/kas.go#L52
as stated in https://github.com/openshift/hypershift/pull/477#issuecomment-932146576?

File: control-plane-operator/controllers/hostedcontrolplane/manifests/nodebootstrapper.go [hunk 16]
```diff
@@ -0,0 +1,58 @@
+package manifests
```
[the remark this answers is not in the document]
Comment (hanqiuzh) [reply control-plane-operator/controllers/hostedcontrolplane/manifests/nodebootstrapper.go]: yes, used that logic and removed the serviceaccount creation 👍

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 17]
```diff
@@ -2785,3 +2794,224 @@ func reconcileMachineConfigServerService(svc *corev1.Service) error {
...
+	cfg := ClusterMachineApproverConfig{
+		NodeClientCert: NodeClientCert{
+			Disabled: false,
+		},
+	}
+	cfgJSON, err := json.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+	b, err := yaml.JSONToYAML(cfgJSON)
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2886]: can't we just yaml.Marshal(cfg)?

Comment (ironcladlou) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2886]: https://github.com/openshift/hypershift/blob/main/api/scheme.go#L29

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 18]
```diff
@@ -2785,3 +2794,224 @@ func reconcileMachineConfigServerService(svc *corev1.Service) error {
...
+func (r *HostedClusterReconciler) reconcileMachineApprover(ctx context.Context, hcluster *hyperv1.HostedCluster, hcp *hyperv1.HostedControlPlane) error {
+	controlPlaneNamespace := manifests.HostedControlPlaneNamespace(hcluster.Namespace, hcluster.Name)
+	err := r.Client.Get(ctx, client.ObjectKeyFromObject(controlPlaneNamespace), controlPlaneNamespace)
+	if err != nil {
+		return fmt.Errorf("failed to get control plane namespace: %w", err)
+	}
+
+	// Reconcile machine-approver role
+	role := machineapprover.Role(controlPlaneNamespace.Name)
+	_, err = controllerutil.CreateOrUpdate(ctx, r.Client, role, func() error {
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2807]: can we update all this calls to createOrUpdate to match this pattern?
```
	tokenSecret := TokenSecret(controlPlaneNamespace, nodePool.Name, targetConfigVersionHash)
	if result, err := controllerutil.CreateOrUpdate(ctx, r.Client, tokenSecret, func() error {
		return reconcileTokenSecret(tokenSecret, nodePool, compressedConfig)
	}); err != nil {
		return ctrl.Result{}, fmt.Errorf("failed to reconcile token Secret: %w", err)
	} else {
		log.Info("Reconciled token Secret", "result", result)
		span.AddEvent("reconciled token Secret", trace.WithAttributes(attribute.String("result", string(result))))
	}
```

Comment (hanqiuzh) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2807]: so inline err handling is preferred? sure, will update 👍

**Comment by enxebre** [comment]
thanks @hanqiuzh! this looks all good to me, just one last ask, since https://github.com/openshift/hypershift/pull/543/files made it in can you please update the createOrUpdate calls to use our new helper instead of controller runtime one as in that PR?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 19]
```diff
@@ -2848,3 +2856,213 @@ func reconcileClusterPrometheusRBAC(ctx context.Context, c client.Client, namesp
...
+		return reconcileMachineApproverRole(role)
+	}); err != nil {
+		return fmt.Errorf("failed to reconcile machine-approver role: %w", err)
+	}
+
+	// Reconcile machine-approver service account
+	sa := machineapprover.ServiceAccount(controlPlaneNamespace.Name)
+	if _, err := r.CreateOrUpdate(ctx, r.Client, sa, NoopReconcile); err != nil {
+		return fmt.Errorf("failed to reconcile machine-approver service account: %w", err)
+	}
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:2878]: I was expecting the pattern with logging and tracing for the operation output, but we can do that later in a different PR to consolidate across the controller.
